fvm/fvm_fuzz_test.go

```diff
@@ -0,0 +1,259 @@
+			txBody := flow.NewTransactionBody().SetScript([]byte("transaction(){prepare(){};execute{panic(\"some panic\")}}"))
+			txBody.SetProposalKey(tctx.address, 0, 0)
+			txBody.SetPayer(tctx.address)
+			return txBody
+		},
+		require: func(t *testing.T, tctx transactionTypeContext, results fuzzResults) {
+			require.Error(t, results.tx.Err)
+			codes := []errors.ErrorCode{
+				errors.ErrCodeComputationLimitExceededError,
+				errors.ErrCodeCadenceRunTimeError, // because of the panic
+			}
+			require.Contains(t, codes, results.tx.Err.Code(), results.tx.Err.Error())
+
+			// fees should be deducted no matter the input
+			fees, deducted := getDeductedFees(t, tctx, results)
+			require.True(t, deducted, "Fees should be deducted.")
+			require.GreaterOrEqual(t, fees.ToGoValue().(uint64), fuzzTestsInclusionFees)
+		},
+	},
+}
+
+const fuzzTestsInclusionFees = uint64(1_000)
+
+// checks fee deduction happened and returns the amount of funds deducted
+func getDeductedFees(tb testing.TB, tctx transactionTypeContext, results fuzzResults) (fees cadence.UFix64, deducted bool) {
+	tb.Helper()
+
+	var ok bool
+	var feesDeductedEvent cadence.Event
+	for _, e := range results.tx.Events {
+		if string(e.Type) == fmt.Sprintf("A.%s.FlowFees.FeesDeducted", fvm.FlowFeesAddress(tctx.chain)) {
+			data, err := jsoncdc.Decode(nil, e.Payload)
+			require.NoError(tb, err)
+			feesDeductedEvent, ok = data.(cadence.Event)
+			require.True(tb, ok, "Event payload should be of type cadence event.")
+		}
+	}
+	if feesDeductedEvent.Type() == nil {
+		return 0, false
+	}
+	fees, ok = feesDeductedEvent.Fields[0].(cadence.UFix64)
+	require.True(tb, ok, "FeesDeducted[0] event should be of type cadence.UFix64.")
+	return fees, true
+}
+
+func bootstrapFuzzStateAndTxContext(tb testing.TB) (bootstrappedVmTest, transactionTypeContext) {
+	tb.Helper()
+
+	addressFunds := uint64(1_000_000_000)
+	var privateKey flow.AccountPrivateKey
+	var address flow.Address
+	bootstrappedVMTest, err := newVMTest().withBootstrapProcedureOptions(
+		fvm.WithTransactionFee(fvm.DefaultTransactionFees),
+		fvm.WithExecutionMemoryLimit(math.MaxUint32),
+		fvm.WithExecutionEffortWeights(mainnetExecutionEffortWeights),
+		fvm.WithExecutionMemoryWeights(meter.DefaultMemoryWeights),
+		fvm.WithMinimumStorageReservation(fvm.DefaultMinimumStorageReservation),
+		fvm.WithAccountCreationFee(fvm.DefaultAccountCreationFee),
+		fvm.WithStorageMBPerFLOW(fvm.DefaultStorageMBPerFLOW),
+	).withContextOptions(
+		fvm.WithTransactionFeesEnabled(true),
+		fvm.WithAccountStorageLimit(true),
+	).bootstrapWith(func(vm *fvm.VirtualMachine, chain flow.Chain, ctx fvm.Context, view state.View, programs *programs.Programs) error {
+		// ==== Create an account ====
+		var txBody *flow.TransactionBody
+		privateKey, txBody = testutil.CreateAccountCreationTransaction(tb, chain)
+
+		err := testutil.SignTransactionAsServiceAccount(txBody, 0, chain)
+		if err != nil {
+			return err
+		}
+
+		tx := fvm.Transaction(txBody, 0)
```

Comment:
can you use incrementing tx index?  you can use `programs.NextTxIndexForTesting()` if you don't want to keep track of the tx index manually (same elsewhere)

Reply:
+1